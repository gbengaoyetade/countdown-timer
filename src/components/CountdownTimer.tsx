import React, { useEffect, useRef } from 'react';


interface CountDownTimerProps {
  time: number;
  timerSpeed: number;
  timerIsStarted: boolean;
  toggleTimerStart: Function;
  setTime: Function;
}

const CountdownTimer = (props: CountDownTimerProps) => {
  const { time, toggleTimerStart, timerSpeed, timerIsStarted, setTime } = props;

  const useInterval = (callback: Function, delay: number) => {
    const savedCallback = useRef<any>(null);

    useEffect(() => {
      savedCallback.current = callback;
    });

    useEffect(
      () => {
        function tick() {
          savedCallback.current();
        }

        const id = setInterval(tick, 1000 / delay);
        return () => clearInterval(id);
      },
      [delay],
    );
  };

  useInterval(() => {
    if (time > 0 && timerIsStarted) {
      setTime(time - 1);
    }
  }, timerSpeed);

  function fmtMSS(s: number) {
    return (s - (s %= 60)) / 60 + (9 < s ? ':' : ':0') + s;
  }

  const getToggleIcon = () => {
    if(!timerIsStarted && time > 0){
      return  <i className="far fa-play-circle"></i>
    }
    return <i className="far fa-pause-circle"></i>
  }
  return (
    <div className="countdown-timer-wrapper">
      <p className="countdown-timer">{fmtMSS(time)}</p>
      <button
        className="toggle-btn"
        onClick={() => time > 0 && toggleTimerStart()} // only works when time is greater than 0
      >
        {getToggleIcon()}
      </button>
    </div>
  );
};

export default CountdownTimer;
