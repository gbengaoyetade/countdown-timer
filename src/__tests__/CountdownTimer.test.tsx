import React from 'react';
import { render } from '@testing-library/react';
import CountdownTimer from '../components/CountdownTimer';


describe('CountdownTimer', () => {
  it('should render without crashing', () => {
    const props = {
      time: 30,
      timerSpeed: 1.5,
      timerIsStarted: false,
      toggleTimerStart: jest.fn(),
      setTime: jest.fn()
    }
    const { container } = render(<CountdownTimer {...props} />);

    expect(container).toMatchSnapshot();
  });

});