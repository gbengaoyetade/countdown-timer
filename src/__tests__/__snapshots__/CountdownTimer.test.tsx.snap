// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownTimer should render without crashing 1`] = `
<div>
  <div
    class="countdown-timer-wrapper"
  >
    <p
      class="countdown-timer"
    >
      0:30
    </p>
    <button
      class="toggle-btn"
    >
      <i
        class="far fa-play-circle"
      />
    </button>
  </div>
</div>
`;
